ebos/hdf5serializer.hh

```diff
@@ -0,0 +1,114 @@
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+#ifndef ECL_HDF5_SERIALIZER_HH
+#define ECL_HDF5_SERIALIZER_HH
+
+#include <opm/common/utility/Serializer.hpp>
+#include <opm/common/utility/MemPacker.hpp>
+
+#include <opm/simulators/utils/HDF5File.hpp>
+#include <opm/simulators/utils/moduleVersion.hpp>
+
+#include <algorithm>
+#include <cctype>
+#include <cstdlib>
+
+namespace Opm {
+
+//! \brief Class for (de-)serializing using HDF5.
+class HDF5Serializer : public Serializer<Serialization::MemPacker> {
+public:
+    HDF5Serializer(const std::string& fileName, HDF5File::OpenMode mode)
+        : Serializer<Serialization::MemPacker>(m_packer_priv)
+        , m_h5file(fileName, mode)
+    {}
+
+    //! \brief Serialize and write data to restart file.
+    //! \tparam T Type of class to write
+    //! \param data Class to write restart data for
+    template<class T>
+    void write(T& data,
+               const std::string& group,
+               const std::string& dset)
+    {
+        try {
+            this->pack(data);
+        } catch (...) {
+            m_packSize = std::numeric_limits<size_t>::max();
+            throw;
+        }
+
+        m_h5file.write(group, dset, m_buffer);
+    }
+
+    //! \brief Writes a header to the file.
+    //! \param simulator_name Name of simulator used
+    //! \param module_version Version of simulator used
+    //! \param time_stamp Build time-stamp for simulator used
+    //! \param case_name Name of case file is associated with
+    void writeHeader(const std::string& simulator_name,
+                     const std::string& module_version,
+                     const std::string& time_stamp,
+                     const std::string& case_name)
+    {
+        try {
+            this->pack(simulator_name, module_version, time_stamp, case_name);
+        } catch (...) {
+            m_packSize = std::numeric_limits<size_t>::max();
+            throw;
+        }
+        m_h5file.write("/", "simulator_info", m_buffer);
+    }
+
+    //! \brief Read data and deserialize from restart file.
+    //! \tparam T Type of class to read
+    //! \param data Class to read restart data for
+    template<class T>
+    void read(T& data,
+              const std::string& group,
+              const std::string& dset)
+    {
+        m_h5file.read(group, dset, m_buffer);
+        this->unpack(data);
+    }
+
+    //! \brief Returns the last report step stored in file.
+    int lastReportStep() const
+    {
+        const auto entries = m_h5file.list("/report_step");
+        std::vector<std::string> filtered;
+        std::copy_if(entries.begin(), entries.end(), std::back_inserter(filtered),
+                     [](const std::string& s)
+                     {
+                         return std::all_of(s.begin(), s.end(), ::isdigit);
+                     });
+        std::sort(filtered.begin(), filtered.end());
```

Comment:
This will sort "9" after "10" so that does not look quite correct to me. Suggest:
```
int last = -1;
for (const auto& entry : entries) {
    int num = std::atoi(entry.c_str()));
    last = std::max(last, num);
}
```

Reply:
arr yes, that was stupid. i'll do the loop.